README.md

[![Travis-CI Build
Status](https://travis-ci.org/wilkox/ggwrap.svg?branch=master)](https://travis-ci.org/wilkox/ggwrap)
[![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/ggwrap)](https://cran.r-project.org/package=ggwrap)

‘ggwrap’ wraps a ‘ggplot2’ plot over multiple rows, to make plots with
long x axes easier to read.

# Installation

The development version of ‘ggwrap’ can be installed from GitHub:

``` r
devtools::install_github("wilkox/ggwrap")
```

# Usage

The `ggwrap` function takes two arguments: a ‘ggplot2’ plot, and the
number of rows over which the plot should be wrapped.

Here’s a simple plot.

``` r
library(ggplot2)
library(ggwrap)
plot <- ggplot(economics, aes(x = date, y = unemploy, colour = uempmed)) +
  geom_line()
plot
```

![](man/figures/README-unnamed-chunk-3-1.png)<!-- -->

We can use `ggwrap` to wrap the plot over four rows.

``` r
ggwrap(plot, 4)
```

![](man/figures/README-unnamed-chunk-4-1.png)<!-- -->


NEWS.md
# v0.0.0.9001

# Major changes

- `ggwrap` now returns a plot object with class 'ggwrap', which can be viewed or saved with `ggsave` (#1, thanks @GuangchuangYu)

# Minor changes

- Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
library(testthat)
library(ggwrap)

test_check("ggwrap")


tests/testthat/test_ggwrap.R
context("ggwrap")

test_that("ggwrap runs without error", {
  plot <- ggplot2::ggplot(
    ggplot2::economics,
    ggplot2::aes(x = date, y = unemploy)
  ) + 
    ggplot2::geom_line()
  expect_silent(ggwrap(plot, 4))
  expect_silent(ggwrap(plot, 40))
})

context("printggwrap")

test_that("print.ggwrap runs without error", {
  plot <- ggplot2::ggplot(
    ggplot2::economics,
    ggplot2::aes(x = date, y = unemploy)
  ) + 
    ggplot2::geom_line()
  plot <- ggwrap(plot, 4)
  expect_silent(plot)
})


R/ggstrip.R
# Make a plot
library(ggplot2)
library(grid)
library(tidyverse)

# Prepare sample data
plot <- economics %>%
  ggplot(aes(x = date, y = unemploy, colour = pop)) +
  geom_point()

ggstrip(plot, 4)

# Given a grob/gtree, return a gtree showing the selected portion of the grob
partial_grob <- function(grob, i, n) {

  # Calculate the number of columns in the grid
  gridcols <- (2 * n) - 1

  # Re-encapsulate the grob with the window as the parent vp
  grob <- gTree(children = gList(grob), vp = "partial::window")

  # Create the gtree
  partial_gtree <- gTree(
    childrenvp = vpTree(
      viewport(
        layout = grid.layout(nrow = 1, ncol = gridcols),
        width = gridcols,
        name = "partial"
      ),
      vpList(viewport(
        layout.pos.col = (n - i + 1):((2 * n) - i),
        layout.pos.row = 1,
        name = "window"
      ))
    ),
    children = gList(grob)
  )

  return(partial_gtree)
}

# Given a ggplot2 plot, return a 'partial'ed strip
partial_strip <- function(plot, i, n) {

  # Turn plot into grob
  pgrob <- ggplotGrob(plot)

  # Extract panel and axes
  panel <- pgrob$grobs[[6]]
  xaxis <- pgrob$grobs[[7]]
  yaxis <- pgrob$grobs[[3]]

  # Take partials of panel and x axis
  panel <- partial_grob(panel, i, n)
  xaxis <- partial_grob(xaxis, i, n)

  # Prepare panel and x axis grobs
  panel_grob <- gTree(children = gList(panel), vp = "strip::panel")
  xaxis_grob <- gTree(children = gList(xaxis), vp = "strip::xaxis")
  yaxis_grob <- gTree(children = gList(yaxis), vp = "strip::yaxis")

  # Prepare gtree for strip
  strip_gtree <- gTree(
    childrenvp = vpTree(
      viewport(layout = grid.layout(
        ncol = 2,
        nrow = 2,
        heights = unit(c(1, 15), c("null", "points")),
        widths = unit(c(5.5, 1), c("point", "null"))
      ) , name = "strip"),
      vpList(
        viewport(layout.pos.row = 1, layout.pos.col = 2, name = "panel", clip = "on"),
        viewport(layout.pos.row = 2, layout.pos.col = 2, name = "xaxis"),
        viewport(layout.pos.row = 1, layout.pos.col = 1, name = "yaxis")
      )
    ),
    children = gList(
      panel_grob,
      xaxis_grob,
      yaxis_grob,
      clipGrob(height = 1, width = 1, vp = "strip::panel")
    )
  )

  return(strip_gtree)
}

ggstrip <- function(plot, nstrips) {

  # Generate strips
  strips <- purrr::map(1:nstrips, ~ partial_strip(plot, .x, nstrips))
  strips <- purrr::map(1:nstrips, ~ gTree(
    children = gList(strips[[.x]]), vp = paste0("strips::strip", .x)
  ))

  # Generate a viewport for each strip
  strip_viewports <- purrr::map(1:nstrips, ~ viewport(
    layout.pos.col = 1,
    layout.pos.row = .x,
    name = paste0("strip", .x)
  ))

  strips_gtree <- gTree(
    childrenvp = vpTree(
      viewport(layout = grid.layout(ncol = 1, nrow = nstrips), name = "strips"),
      do.call(vpList, strip_viewports)
    ),
    children = do.call(gList, strips)
  )

  # Turn plot into a grob
  pgrob <- ggplotGrob(plot)

  # Replace panel area with strips
  pgrob$grobs[[6]] <- strips_gtree

  # Remove axes
  pgrob$grobs[[3]] <- zeroGrob()
  pgrob$grobs[[7]] <- zeroGrob()

  grid.draw(pgrob)
}


R/ggwrap.R
#' Wrap a 'gglot2' plot over multiple rows.
#'
#' @return Returns a plot object which can be viewed or saved to file with
#' `ggplot2::ggsave`.
#'
#' @param plot A 'ggplot2' plot to be wrapped.
#' @param n The number of rows in the wrapped plot.
#'
#' @examples
#' plot <- ggplot2::ggplot(
#'   ggplot2::economics,
#'   ggplot2::aes(x = date, y = unemploy, colour = uempmed)
#' ) +
#'   ggplot2::geom_line()
#' plot <- ggwrap(plot, 4)
#' @export
ggwrap <- function(plot, n) {

  # Generate strips
  strips <- purrr::map(1:n, ~ partial_strip(plot, .x, n))
  strips <- purrr::map(1:n, ~ grid::gTree(
    children = grid::gList(strips[[.x]]), vp = paste0("strips::strip", .x)
  ))

  # Generate a viewport for each strip
  strip_viewports <- purrr::map(1:n, ~ grid::viewport(
    layout.pos.col = 1,
    layout.pos.row = .x,
    name = paste0("strip", .x)
  ))

  strips_gtree <- grid::gTree(
    childrenvp = grid::vpTree(
      grid::viewport(
        layout = grid::grid.layout(ncol = 1, nrow = n),
        name = "strips"
      ),
      do.call(grid::vpList, strip_viewports)
    ),
    children = do.call(grid::gList, strips)
  )

  # Turn plot into a grob
  pgrob <- ggplot2::ggplotGrob(plot)

  # Replace panel area with strips
  pgrob$grobs[[6]] <- strips_gtree

  # Remove axes
  pgrob$grobs[[3]] <- ggplot2::zeroGrob()
  pgrob$grobs[[7]] <- ggplot2::zeroGrob()

  # Add 'ggwrap' class
  class(pgrob) <- c("ggwrap", class(pgrob))

  # Return
  pgrob
}

#' @export
print.ggwrap <- function(x, ...) {

  grid::grid.newpage()

  # Record dependency on 'ggplot2' on the display list
  # (AFTER grid.newpage())
  grDevices::recordGraphics(
    requireNamespace("ggplot2", quietly = TRUE),
    list(),
    getNamespace("ggplot2")
  )

  grid::grid.draw(x)

  invisible(x)
}

# Select a horizontal slice of a grob. Not exported.
partial_grob <- function(grob, i, n) {

  # Calculate the number of columns in the grid
  gridcols <- (2 * n) - 1

  # Re-encapsulate the grob with the window as the parent vp
  grob <- grid::gTree(
    children = grid::gList(grob),
    vp = "partial::window"
  )

  # Create the gtree
  partial_gtree <- grid::gTree(
    childrenvp = grid::vpTree(
      grid::viewport(
        layout = grid::grid.layout(nrow = 1, ncol = gridcols),
        width = gridcols,
        name = "partial"
      ),
      grid::vpList(grid::viewport(
        layout.pos.col = (n - i + 1):((2 * n) - i),
        layout.pos.row = 1,
        name = "window"
      ))
    ),
    children = grid::gList(grob)
  )

  return(partial_gtree)
}

# Draw a 'strip' of a 'ggplot2' plot containing only a horizontal slice of the
# plot. Not exported.
partial_strip <- function(plot, i, n) {

  # Turn plot into grob
  pgrob <- ggplot2::ggplotGrob(plot)

  # Extract panel and axes
  panel <- pgrob$grobs[[6]]
  xaxis <- pgrob$grobs[[7]]
  yaxis <- pgrob$grobs[[3]]

  # The x axis TableGrob sneakily tries to tell the x axis not to clip to its
  # parent viewport
  xaxis$children$axis$layout$clip <- c("inherit", "inherit")

  # Take partials of panel and x axis
  panel <- partial_grob(panel, i, n)
  xaxis <- partial_grob(xaxis, i, n)

  # Prepare panel and x axis grobs
  panel_grob <- grid::gTree(children = grid::gList(panel), vp = "strip::panel")
  xaxis_grob <- grid::gTree(children = grid::gList(xaxis), vp = "strip::xaxis")
  yaxis_grob <- grid::gTree(children = grid::gList(yaxis), vp = "strip::yaxis")

  # Prepare gtree for strip
  strip_gtree <- grid::gTree(
    childrenvp = grid::vpTree(
      grid::viewport(layout = grid::grid.layout(
        ncol = 2,
        nrow = 2,
        heights = grid::unit(c(1, 15), c("null", "points")),
        widths = grid::unit(c(5.5, 1), c("point", "null"))
      ) , name = "strip"),
      grid::vpList(
        grid::viewport(
          layout.pos.row = 1,
          layout.pos.col = 2,
          name = "panel",
          clip = "on"
        ),
        grid::viewport(
          layout.pos.row = 2,
          layout.pos.col = 2,
          name = "xaxis",
          clip = "on"
        ),
        grid::viewport(
          layout.pos.row = 1,
          layout.pos.col = 1,
          name = "yaxis"
        )
      )
    ),
    children = grid::gList(
      panel_grob,
      xaxis_grob,
      yaxis_grob
    )
  )

  return(strip_gtree)
}

